src/agent.ts

```diff
@@ -108,6 +113,39 @@ export class Agent {
       history = modified.history;
     }
 
+    let hasMoreActions = true;
+    let iter = 0;
+    do {
+      const inference = await this.performInference(
```

Comment:
How do lifecycles work here?  With looping, this breaks lifecycles.  The "before call" and "before inference" calls are no longer doing what we expect.